File: src/domain/budgets.ts
// Budgets (SPEC §8.1.6, no rollover in Phase 1). CONTRACT — implemented by
// the domain build agent.
//
// Semantics:
//  * a budget covers its categoryIds PLUS all their descendants (D16);
//  * spend counts expense outflows net of refunds: spentMinor =
//    -(Σ amounts in those categories within the window), floored at
//    whatever the maths says (can go negative if refunds exceed spend);
//  * split transactions contribute each split by the SPLIT's category;
//  * transfers never count (their categoryId is null);
//  * amounts are converted to the base currency at current rates; transactions
//    with no available rate are counted in missingRateCount, never guessed
//    (SPEC §6, D22);
//  * period windows anchor at startDate: weekly = consecutive 7-day windows,
//    monthly = calendar-ish months anchored at startDate's day (clamped to
//    short months), yearly = anniversary years. Windows are [start, end]
//    inclusive of both dates.
import dayjs from 'dayjs';
import { db, getSettings } from '../db/db';
import type { Budget, BudgetPeriod, Transaction } from '../db/types';
import { todayISO, uid } from '../lib/util';
import { convertMinor } from '../money/money';
import { descendantIds } from './categories';
import { rateLookup } from './fx';
import { ValidationError } from './transactions';

export interface PeriodWindow {
  start: string; // 'YYYY-MM-DD' inclusive
  end: string; // 'YYYY-MM-DD' inclusive
}

const DAY_MS = 86_400_000;
const ISO_FMT = 'YYYY-MM-DD';

/** Exact whole-day timestamp for a 'YYYY-MM-DD' string (UTC → no DST drift). */
const utcMs = (isoDate: string): number => Date.parse(`${isoDate}T00:00:00Z`);

/**
 * The n-th window of a budget's period grid (n 0 = the window starting at
 * startDate; n may be negative). Monthly/yearly windows are
 * [anchor.add(n, unit), anchor.add(n+1, unit) - 1 day] — dayjs clamps a
 * short month's end (31 Jan + 1 month = 28/29 Feb), and because both ends
 * derive from the SAME anchor the clamping never accumulates and windows tile
 * the timeline contiguously with no gaps or overlaps.
 */
function windowAt(budget: Pick<Budget, 'period' | 'startDate'>, n: number): PeriodWindow {
  const anchor = dayjs(budget.startDate);
  if (budget.period === 'weekly') {
    const s = anchor.add(n * 7, 'day');
    return { start: s.format(ISO_FMT), end: s.add(6, 'day').format(ISO_FMT) };
  }
  const unit = budget.period === 'monthly' ? 'month' : 'year';
  return {
    start: anchor.add(n, unit).format(ISO_FMT),
    end: anchor
      .add(n + 1, unit)
      .subtract(1, 'day')
      .format(ISO_FMT),
  };
}

/** Index n of the window containing `date` (negative before startDate). */
function windowIndexOf(budget: Pick<Budget, 'period' | 'startDate'>, date: string): number {
  if (budget.period === 'weekly') {
    // floor() handles dates before startDate: -1 day → window -1, etc.
    return Math.floor((utcMs(date) - utcMs(budget.startDate)) / (7 * DAY_MS));
  }
  const [sy, sm] = budget.startDate.split('-').map(Number);
  const [dy, dm] = date.split('-').map(Number);
  let n = budget.period === 'monthly' ? (dy - sy) * 12 + (dm - sm) : dy - sy;
  // The year/month guess can be off by one around clamped short-month ends;
  // windows tile contiguously, so walk (at most a step or two) to the one
  // containing `date`. 'YYYY-MM-DD' strings compare correctly lexicographically.
  while (date < windowAt(budget, n).start) n -= 1;
  while (date > windowAt(budget, n).end) n += 1;
  return n;
}

/** Pure, unit-tested. Window of the budget's period containing `date`. */
export function windowContaining(
  budget: Pick<Budget, 'period' | 'startDate'>,
  date: string,
): PeriodWindow {
  return windowAt(budget, windowIndexOf(budget, date));
}

/**
 * Pure. Shift a window by n periods (n may be negative). The window's index on
 * the budget's grid is recovered from its start date, so shifting stays exact
 * even where clamping makes naive start/end ± n arithmetic drift
 * (e.g. monthly anchored on the 31st crossing February).
 */
export function shiftWindow(
  budget: Pick<Budget, 'period' | 'startDate'>,
  window: PeriodWindow,
  n: number,
): PeriodWindow {
  return windowAt(budget, windowIndexOf(budget, window.start) + n);
}

export interface BudgetProgress {
  budget: Budget;
  window: PeriodWindow;
  spentMinor: number; // net spend as a positive number (refunds subtract)
  limitMinor: number;
  remainingMinor: number; // limit - spent; negative when over
  pct: number; // spent / limit; may exceed 1; 0 when limit is 0
  over: boolean;
  missingRateCount: number;
}

/**
 * The signed minor-unit contributions of one transaction to a budget whose
 * covered-category set is `cats`:
 *  * transfer legs contribute nothing (D13);
 *  * a split transaction contributes ONLY its splits whose category is
 *    covered (the parent categoryId is ignored when splits exist);
 *  * otherwise the whole amount when the transaction's category is covered.
 * Each contribution is converted (and rounded) individually, matching how
 * reports attribute split lines to their own categories.
 */
function contributionsOf(t: Transaction, cats: Set<string>): number[] {
  if (t.transferGroupId !== null) return [];
  if (t.splits.length > 0) {
    return t.splits
      .filter((s) => s.categoryId !== null && cats.has(s.categoryId))
      .map((s) => s.amountMinor);
  }
  return t.categoryId !== null && cats.has(t.categoryId) ? [t.amountMinor] : [];
}

export async function budgetProgress(budget: Budget, refDate?: string): Promise<BudgetProgress> {
  const window = windowContaining(budget, refDate ?? todayISO());
  const [allCats, settings, lookup, txs] = await Promise.all([
    db.categories.toArray(),
    getSettings(),
    rateLookup(),
    db.transactions.where('date').between(window.start, window.end, true, true).toArray(),
  ]);
  const cats = descendantIds(allCats, budget.categoryIds);

  let sumMinor = 0; // signed Σ of converted contributions (spend negative)
  let missingRateCount = 0;
  for (const t of txs) {
    for (const amountMinor of contributionsOf(t, cats)) {
      const converted = convertMinor(amountMinor, t.currency, settings.baseCurrency, lookup);
      if (converted === null) missingRateCount += 1; // never guess (SPEC §6)
      else sumMinor += converted;
    }
  }

  // Expenses are negative → spend is positive. Guard against IEEE -0 so a
  // zero-spend window can never display as "-£0.00".
  const spentMinor = sumMinor === 0 ? 0 : -sumMinor;
  const limitMinor = budget.amountMinor;
  return {
    budget,
    window,
    spentMinor,
    limitMinor,
    remainingMinor: limitMinor - spentMinor,
    pct: limitMinor > 0 ? Math.max(0, spentMinor / limitMinor) : 0,
    over: spentMinor > limitMinor,
    missingRateCount,
  };
}

/** Progress for all non-archived budgets, sorted by name. */
export async function allBudgetProgress(refDate?: string): Promise<BudgetProgress[]> {
  // NB: `archived` is a boolean — booleans are not valid IndexedDB keys, so
  // the `archived` index cannot be queried with equals(); filter in memory.
  const budgets = (await db.budgets.toArray())
    .filter((b) => !b.archived)
    .sort((a, b) => a.name.localeCompare(b.name));
  return Promise.all(budgets.map((b) => budgetProgress(b, refDate)));
}

export interface SaveBudgetInput {
  id?: string;
  name: string;
  categoryIds: string[];
  amountMinor: number; // base currency (D22)
  period: BudgetPeriod;
  startDate: string;
  archived?: boolean;
}

const PERIODS: readonly BudgetPeriod[] = ['weekly', 'monthly', 'yearly'];

/** True for a real 'YYYY-MM-DD' calendar date (rejects e.g. 2025-02-30). */
function isValidISODate(s: string): boolean {
  return /^\d{4}-\d{2}-\d{2}$/.test(s) && dayjs(s).format(ISO_FMT) === s;
}

export async function saveBudget(input: SaveBudgetInput): Promise<Budget> {
  const name = input.name.trim().replace(/\s+/g, ' ');
  if (!name) throw new ValidationError('Budget name cannot be empty');
  if (!Number.isSafeInteger(input.amountMinor) || input.amountMinor <= 0) {
    throw new ValidationError('Budget amount must be a positive amount');
  }
  const categoryIds = [...new Set(input.categoryIds)].filter(Boolean);
  if (categoryIds.length === 0) {
    throw new ValidationError('Choose at least one category');
  }
  if (!PERIODS.includes(input.period)) {
    throw new ValidationError('Budget period must be weekly, monthly or yearly');
  }
  if (!isValidISODate(input.startDate)) {
    throw new ValidationError('Start date must be a valid YYYY-MM-DD date');
  }
  return db.transaction('rw', db.budgets, async () => {
    const existing = input.id ? await db.budgets.get(input.id) : undefined;
    if (input.id && !existing) throw new ValidationError('Budget not found');
    const budget: Budget = {
      id: existing?.id ?? uid(),
      name,
      categoryIds,
      amountMinor: input.amountMinor,
      period: input.period,
      startDate: input.startDate,
      rollover: existing?.rollover ?? false, // Phase 2; preserved, not editable yet
      archived: input.archived ?? existing?.archived ?? false,
    };
    await db.budgets.put(budget);
    return budget;
  });
}

export async function deleteBudget(id: string): Promise<void> {
  // Nothing references budgets, so a hard delete is safe.
  await db.budgets.delete(id);
}

File: native/Sources/MyMoneyKit/Domain/Budgets.swift
// Budget period windows and spend against them, ported from
// src/domain/budgets.ts (SPEC §8.1.6, no rollover in Phase 1).
//
// THE WINDOW GRID IS THE WHOLE PROBLEM. A budget anchors at its `startDate`
// and its windows tile the timeline forwards AND backwards from there, with no
// gaps and no overlaps -- every date is in exactly one window, including dates
// before the anchor (negative window indices). Window n is
//
//     [ anchor + n periods , anchor + (n+1) periods − 1 day ]
//
// with BOTH ends computed from the same anchor. That detail is load-bearing:
// a monthly budget anchored on the 31st has its end clamped every February
// (31 Jan + 1 month = 28 Feb, so window 0 ends on the 27th... no: on the day
// before, 28 Feb), and because the next window's START is also computed from
// the anchor rather than from the clamped end, the clamping never accumulates.
// Deriving each window from the previous one instead -- the obvious
// implementation -- drifts a day earlier every short month until the budget no
// longer means "the 31st", and the oracle's `budgets.window.monthly.31st-*`
// cases exist to catch exactly that.
//
// SPEND IS POSITIVE, and net of refunds. Expenses are stored negative, so
// spend = −Σ; a refund is a positive amount in an expense category and it
// SUBTRACTS. A month whose refunds exceed its spending has a negative
// `spentMinor`, and that is reported as-is rather than floored at zero,
// because the money really did come back.
//
// A MISSING RATE IS COUNTED, NEVER GUESSED, and counted ONCE PER TRANSACTION
// however many of that transaction's splits the budget covers (D28). The UI
// says "N transactions excluded"; counting per split would over-report a split
// purchase and make the number mean something other than what it says.
import Foundation

public enum DomainError: Error, Equatable, Sendable, CustomStringConvertible {
    /// A date string that is not a real 'YYYY-MM-DD' calendar date. A refusal:
    /// the alternative is picking a nearby day, which moves someone's money
    /// into a period they did not spend it in.
    case invalidDate(String)

    public var description: String {
        switch self {
        case .invalidDate(let s): return "\"\(s)\" is not a YYYY-MM-DD calendar date"
        }
    }
}

/// A budget period window, INCLUSIVE of both ends.
public struct PeriodWindow: Sendable, Hashable {
    public let start: String  // 'YYYY-MM-DD'
    public let end: String    // 'YYYY-MM-DD'

    public init(start: String, end: String) {
        self.start = start
        self.end = end
    }

    /// True when `date` ('YYYY-MM-DD') falls inside. String comparison is date
    /// comparison for this format, which is why the format was chosen.
    public func contains(_ date: String) -> Bool { date >= start && date <= end }
}

/// The part of a budget the window grid and the spend calculation actually
/// use. Separate from `Budget` because the oracle's cases state exactly these
/// fields and nothing else -- a function that demanded an id and a name would
/// force the harness to invent two.
public struct BudgetSpec: Sendable, Hashable {
    public let categoryIds: [String]
    /// In BASE currency (D22).
    public let amountMinor: Int64
    public let period: BudgetPeriod
    public let startDate: String

    public init(categoryIds: [String], amountMinor: Int64, period: BudgetPeriod, startDate: String) {
        self.categoryIds = categoryIds
        self.amountMinor = amountMinor
        self.period = period
        self.startDate = startDate
    }
}

extension Budget {
    public var spec: BudgetSpec {
        BudgetSpec(
            categoryIds: categoryIds, amountMinor: amountMinor,
            period: period, startDate: startDate
        )
    }
}

public struct BudgetProgress: Sendable, Hashable {
    public let window: PeriodWindow
    /// Net spend as a POSITIVE number; negative when refunds exceeded spend.
    public let spentMinor: Int64
    public let limitMinor: Int64
    /// limit − spent; negative when over.
    public let remainingMinor: Int64
    /// spent / limit. Floored at 0, NOT capped at 1, and 0 when the limit is 0.
    /// The one Double in the money layer, and it is a ratio for a progress bar,
    /// never an amount.
    public let pct: Double
    /// spent > limit. Spending EXACTLY the limit is not over.
    public let over: Bool
    /// Transactions left out because their currency has no rate to base.
    public let missingRateCount: Int
}

public enum Budgets {
    // MARK: - The window grid

    /// Window `n` of the grid; `n` may be negative.
    static func window(_ period: BudgetPeriod, anchor: CalendarDate, index n: Int) -> PeriodWindow {
        switch period {
        case .weekly:
            let start = anchor.addingDays(n * 7)
            return PeriodWindow(start: start.iso, end: start.addingDays(6).iso)
        case .monthly:
            return PeriodWindow(
                start: anchor.addingMonths(n).iso,
                end: anchor.addingMonths(n + 1).addingDays(-1).iso
            )
        case .yearly:
            return PeriodWindow(
                start: anchor.addingYears(n).iso,
                end: anchor.addingYears(n + 1).addingDays(-1).iso
            )
        }
    }

    /// Index of the window containing `date`.
    ///
    /// For weeks it is exact arithmetic. For months and years the year/month
    /// difference is a GUESS that can be off by one around a clamped
    /// short-month end, so the guess is corrected by walking -- at most a step
    /// or two, because the windows tile contiguously. The walk is the
    /// TypeScript's, kept rather than replaced with something cleverer: a
    /// closed form for "which clamped anniversary window contains this date"
    /// is easy to get subtly wrong and impossible to read.
    static func windowIndex(
        _ period: BudgetPeriod, anchor: CalendarDate, date: CalendarDate
    ) -> Int {
        if period == .weekly {
            // Floored, not truncated: a date one day BEFORE the anchor is in
            // window −1, and Swift's `/` would put it in window 0.
            let diff = date.daysSince(anchor)
            return diff >= 0 ? diff / 7 : -(((-diff) + 6) / 7)
        }
        var n = period == .monthly
            ? (date.year - anchor.year) * 12 + (date.month - anchor.month)
            : date.year - anchor.year
        while date.iso < window(period, anchor: anchor, index: n).start { n -= 1 }
        while date.iso > window(period, anchor: anchor, index: n).end { n += 1 }
        return n
    }

    /// The window of this budget's period containing `date`.
    public static func windowContaining(
        period: BudgetPeriod, startDate: String, date: String
    ) throws -> PeriodWindow {
        guard let anchor = CalendarDate(iso: startDate) else { throw DomainError.invalidDate(startDate) }
        guard let d = CalendarDate(iso: date) else { throw DomainError.invalidDate(date) }
        return window(period, anchor: anchor, index: windowIndex(period, anchor: anchor, date: d))
    }

    /// Shift a window by `n` periods (`n` may be negative).
    ///
    /// The window's INDEX on the grid is recovered from its start date and the
    /// shift is applied to the index -- not to the dates. Adding n months to
    /// both ends of a clamped window drifts (28 Feb + 1 month = 28 Mar, but
    /// the real next window starts on the 31st), and the drift compounds.
    public static func shiftWindow(
        period: BudgetPeriod, startDate: String, window w: PeriodWindow, by n: Int
    ) throws -> PeriodWindow {
        guard let anchor = CalendarDate(iso: startDate) else { throw DomainError.invalidDate(startDate) }
        guard let start = CalendarDate(iso: w.start) else { throw DomainError.invalidDate(w.start) }
        return window(period, anchor: anchor, index: windowIndex(period, anchor: anchor, date: start) + n)
    }

    // MARK: - Spend against a window

    /// The signed contributions of one transaction to a budget covering `cats`:
    ///  * a transfer leg contributes nothing (D13) -- moving money between your
    ///    own accounts is not spending, and counting it would double every
    ///    savings transfer;
    ///  * a SPLIT transaction contributes only the splits whose category is
    ///    covered, and the parent's own categoryId is ignored entirely;
    ///  * otherwise the whole amount, when the transaction's category is
    ///    covered.
    ///
    /// Public because the budget DETAIL screen lists the transactions behind
    /// the figure, and that list has to be decided by the same predicate the
    /// total is -- otherwise the rows on screen would not add up to the number
    /// above them, which is the most alarming thing a budget screen can do.
    public static func contributions(of tx: Transaction, covering cats: Set<String>) -> [Int64] {
        if tx.transferGroupId != nil { return [] }
        if !tx.splits.isEmpty {
            return tx.splits.compactMap { split in
                guard let id = split.categoryId, cats.contains(id) else { return nil }
                return split.amountMinor
            }
        }
        guard let id = tx.categoryId, cats.contains(id) else { return [] }
        return [tx.amountMinor]
    }

    /// Progress for one budget as of `refDate`.
    ///
    /// Each contribution is converted and rounded INDIVIDUALLY, matching how
    /// reports attribute split lines to their own categories -- and matching
    /// `fx.convertEach`, which pins that converting three amounts and adding
    /// them is not the same as adding and converting once.
    public static func progress(
        _ budget: BudgetSpec,
        categories: [Category],
        transactions: [Transaction],
        baseCurrency: String,
        rates: RateTable,
        refDate: String
    ) throws -> BudgetProgress {
        let w = try windowContaining(
            period: budget.period, startDate: budget.startDate, date: refDate
        )
        let cats = Categories.descendantIds(categories, rootIds: budget.categoryIds)

        var sumMinor: Int64 = 0  // signed Σ of converted contributions (spend negative)
        var missingRateCount = 0
        for tx in transactions where w.contains(tx.date) {
            let parts = contributions(of: tx, covering: cats)
            if parts.isEmpty { continue }
            // Convertibility is a property of the TRANSACTION's currency, so an
            // unconvertible transaction is excluded -- and counted -- exactly
            // once however many of its splits this budget covers (D28).
            if rates.rate(from: tx.currency, to: baseCurrency) == nil {
                missingRateCount += 1
                continue
            }
            for amount in parts {
                switch Money.convert(minor: amount, from: tx.currency, to: baseCurrency, using: rates) {
                case .converted(let value):
                    let (next, overflowed) = sumMinor.addingReportingOverflow(value)
                    if overflowed { throw MoneyError.overflow("budget spend total") }
                    sumMinor = next
                case .missingRate:
                    // Unreachable: the rate was checked above. Left as a case
                    // rather than a `try!` so a future change cannot make it a
                    // silent zero.
                    missingRateCount += 1
                case .notRepresentable:
                    throw MoneyError.notRepresentable("a contribution to this budget")
                }
            }
        }

        // Expenses are negative, so spend is the negation. Checked, not `-x`:
        // the one input that traps is Int64.min, and a trap in a budget screen
        // is a crash where a refusal belongs.
        let (spent, negated) = (0 as Int64).subtractingReportingOverflow(sumMinor)
        if negated { throw MoneyError.overflow("budget spend total") }
        let limit = budget.amountMinor
        let (remaining, remainingOverflowed) = limit.subtractingReportingOverflow(spent)
        if remainingOverflowed { throw MoneyError.overflow("budget remaining") }
        return BudgetProgress(
            window: w,
            spentMinor: spent,
            limitMinor: limit,
            remainingMinor: remaining,
            pct: limit > 0 ? max(0, Double(spent) / Double(limit)) : 0,
            over: spent > limit,
            missingRateCount: missingRateCount
        )
    }
}

extension Book {
    /// Progress for a budget against this book.
    public func budgetProgress(_ budget: BudgetSpec, refDate: String) throws -> BudgetProgress {
        try Budgets.progress(
            budget,
            categories: categories,
            transactions: transactions,
            baseCurrency: baseCurrency,
            rates: rateTable,
            refDate: refDate
        )
    }
}
